ensime_shared/ensime.py

```diff
@@ -300,3 +362,13 @@ def fun_en_complete_func(self, client, findstart_and_base, base=None):
             # Invoked by vim
             findstart = findstart_and_base
         return client.complete_func(findstart, base)
+
+    def watchdog_client_server_connection(self, client):
+        """Run periodically to check and maintain client, server, and cxn health."""
+        if not client.connected and client.connection_attempts < 10:
+            config = client.config
+            server = self.server(config)
+            if not server.isrunning():
+                self.start_server(config)
+            client.connect(server)
+            client.connection_attempts += 1
```

Comment:
I'm just preserving how the prior implementation behaved so far, but while we're looking at this, it seems excessive to me to check this on every `CursorMove`—`CursorHold` is probably fine, but `CursorMove` a bit much. The server and our connection are not so unstable. Might experiment with removing the call to this from `au_cursor_moved`.